support/cas-server-support-oauth-core-api/src/main/java/org/apereo/cas/support/oauth/authenticator/OAuth20ProofKeyCodeExchangeAuthenticator.java

```diff
@@ -29,20 +30,34 @@
  */
 @Slf4j
 public class OAuth20ProofKeyCodeExchangeAuthenticator extends OAuth20ClientIdClientSecretAuthenticator {
-    private final TicketRegistry ticketRegistry;
 
     public OAuth20ProofKeyCodeExchangeAuthenticator(final ServicesManager servicesManager,
                                                     final ServiceFactory webApplicationServiceFactory,
                                                     final AuditableExecution registeredServiceAccessStrategyEnforcer,
                                                     final TicketRegistry ticketRegistry,
                                                     final CipherExecutor<Serializable, String> registeredServiceCipherExecutor) {
-        super(servicesManager, webApplicationServiceFactory, registeredServiceAccessStrategyEnforcer, registeredServiceCipherExecutor);
-        this.ticketRegistry = ticketRegistry;
+        super(servicesManager, webApplicationServiceFactory, registeredServiceAccessStrategyEnforcer, registeredServiceCipherExecutor, ticketRegistry);
+    }
+
+    /**
+     * Can perform authentication if the code_verifier is supplied.
+     * @param context the context
+     * @return
+     */
+    @Override
+    protected boolean canAuthenticate(final WebContext context) {
+        return !context.getRequestParameter(OAuth20Constants.CODE_VERIFIER).isEmpty();
     }
 
     @Override
     protected void validateCredentials(final UsernamePasswordCredentials credentials,
                                        final OAuthRegisteredService registeredService, final WebContext context) {
+        val clientSecret = context.getRequestParameter(OAuth20Constants.CLIENT_SECRET).map(String::valueOf).orElse(StringUtils.EMPTY);
+
+        if (!OAuth20Utils.checkClientSecret(registeredService, clientSecret, registeredServiceCipherExecutor)) {
+            throw new CredentialsException("Bad secret for client identifier: " + credentials.getUsername());
```

Comment:
Modify the exception message to make the error message more generic; something like "Unable to verify client based on service X in the registry". Then push a message to the ERROR log explaining exactly what happened.

Reply:
I tried keeping the error handling same as in https://github.com/apereo/cas/blob/master/support/cas-server-support-oauth-core-api/src/main/java/org/apereo/cas/support/oauth/authenticator/OAuth20UsernamePasswordAuthenticator.java#L56 and 
https://github.com/apereo/cas/blob/master/support/cas-server-support-oauth-core-api/src/main/java/org/apereo/cas/support/oauth/authenticator/OAuth20ClientIdClientSecretAuthenticator.java#L83

Reply:
Great. Let's make sure that spot is also cleaned up as well.

Reply:
Fixed the exception message in all places. On logging error, its already done at https://github.com/apereo/cas/blob/master/support/cas-server-support-oauth-core-api/src/main/java/org/apereo/cas/support/oauth/util/OAuth20Utils.java#L418